File: processing/models/edit_config.py
from pydantic import BaseModel
from typing import List, Optional


# AI Analysis models

class Hook(BaseModel):
    start: float
    end: float
    text: str
    score: float


class DeadMoment(BaseModel):
    start: float
    end: float
    reason: str  # "silence" | "uhm" | "repetition" | "filler"


class KeyMoment(BaseModel):
    time: float
    type: str  # "emotional_peak" | "key_insight" | "humor" | "surprise"
    description: str
    suggestedZoomScale: float = 1.15
    highlightWords: List[str] = []


class SuggestedClip(BaseModel):
    start: float
    end: float
    title: str
    hookScore: float
    viralityEstimate: str  # "low" | "medium" | "high"
    reason: str
    suggestedHookReorder: Optional[dict] = None


class TopicSegment(BaseModel):
    start: float
    end: float
    topic: str


class AiAnalysis(BaseModel):
    summary: str
    hooks: List[Hook]
    deadMoments: List[DeadMoment]
    keyMoments: List[KeyMoment]
    suggestedClips: List[SuggestedClip]
    topicSegments: List[TopicSegment]


# Edit Config models

class CutConfig(BaseModel):
    id: str
    start: float
    end: float
    reason: str


class ZoomConfig(BaseModel):
    id: str
    time: float
    duration: float
    scale: float
    easing: str = "ease_in_out"
    anchorX: float = 0.5
    anchorY: float = 0.5
    reason: str = ""


class ReframingConfig(BaseModel):
    enabled: bool = True
    mode: str = "face_track"
    manualCropX: Optional[float] = None


class CaptionConfig(BaseModel):
    enabled: bool = True
    style: str = "hormozi"
    position: str = "bottom"
    fontSize: str = "medium"
    primaryColor: str = "#FFFFFF"
    highlightColor: str = "#FFD700"
    backgroundColor: Optional[str] = None
    font: str = "Inter"
    maxWordsPerLine: int = 4
    animation: str = "word_by_word"
    highlightKeywords: bool = True
    customKeywords: List[str] = []


class TransitionConfig(BaseModel):
    intro: str = "fade_in"
    outro: str = "none"
    betweenCuts: str = "hard"


class MusicConfig(BaseModel):
    enabled: bool = False
    track: Optional[str] = None
    volume: float = 0.1
    duckOnSpeech: bool = True


class SoundEffectsConfig(BaseModel):
    enabled: bool = False
    whooshOnCut: bool = False
    boomOnKeyMoment: bool = False
    volume: float = 0.3


class AudioConfig(BaseModel):
    normalizeVolume: bool = True
    removeBackgroundNoise: bool = False
    music: MusicConfig = MusicConfig()
    soundEffects: SoundEffectsConfig = SoundEffectsConfig()


class OverlayConfig(BaseModel):
    progressBar: bool = False
    hookText: Optional[str] = None
    ctaText: Optional[str] = None
    watermark: dict = {"enabled": False, "imageUrl": None, "position": "bottom_right"}


class SegmentConfig(BaseModel):
    id: str
    sourceStart: float
    sourceEnd: float
    transition: str = "none"  # "none"|"hard"|"crossfade"|"fade"|"wipe_left"|"wipe_right"|"slide_up"|"dissolve"|"zoom_in"|"circle"
    transitionDuration: float = 0.3  # seconds


class CaptionOverride(BaseModel):
    text: Optional[str] = None
    hidden: bool = False
    highlight: bool = False


class EditConfig(BaseModel):
    outputRatio: str = "9:16"
    segments: List[SegmentConfig] = []
    cuts: List[CutConfig] = []
    zooms: List[ZoomConfig] = []
    reframing: ReframingConfig = ReframingConfig()
    captions: CaptionConfig = CaptionConfig()
    transitions: TransitionConfig = TransitionConfig()
    audio: AudioConfig = AudioConfig()
    overlays: OverlayConfig = OverlayConfig()
    captionOverrides: dict = {}  # {wordIndex: CaptionOverride}

File: apps/web/src/types/editConfig.ts
export type OutputRatio = "9:16" | "1:1" | "4:5";

export interface CutConfig {
  id: string;
  start: number;
  end: number;
  reason: "silence" | "uhm" | "repetition" | "filler" | "manual";
}

export interface ZoomConfig {
  id: string;
  time: number;
  duration: number;
  scale: number;
  easing: "ease_in_out" | "ease_in" | "linear" | "snap";
  anchorX: number;
  anchorY: number;
  reason: string;
}

export interface ReframingConfig {
  enabled: boolean;
  mode: "face_track" | "center" | "manual";
  manualCropX: number | null;
}

export type CaptionStyleType =
  | "hormozi"
  | "minimal"
  | "karaoke"
  | "bold"
  | "outline"
  | "custom";

export interface CaptionConfig {
  enabled: boolean;
  style: CaptionStyleType;
  position: "bottom" | "center" | "top";
  fontSize: "small" | "medium" | "large";
  primaryColor: string;
  highlightColor: string;
  backgroundColor: string | null;
  font: string;
  maxWordsPerLine: number;
  animation: "word_by_word" | "line_by_line" | "fade" | "none";
  highlightKeywords: boolean;
  customKeywords: string[];
}

export interface TransitionConfig {
  intro: "none" | "fade_in" | "slide_up" | "zoom_in";
  outro: "none" | "fade_out" | "slide_down";
  betweenCuts: "hard" | "crossfade" | "swipe";
}

export interface MusicConfig {
  enabled: boolean;
  track: string | null;
  volume: number;
  duckOnSpeech: boolean;
}

export interface SoundEffectsConfig {
  enabled: boolean;
  whooshOnCut: boolean;
  boomOnKeyMoment: boolean;
  volume: number;
}

export interface AudioConfig {
  normalizeVolume: boolean;
  removeBackgroundNoise: boolean;
  music: MusicConfig;
  soundEffects: SoundEffectsConfig;
}

export interface OverlayConfig {
  progressBar: boolean;
  hookText: string | null;
  ctaText: string | null;
  watermark: {
    enabled: boolean;
    imageUrl: string | null;
    position: string;
  };
}

export type SegmentTransition =
  | "none"
  | "hard"
  | "crossfade"
  | "fade"
  | "wipe_left"
  | "wipe_right"
  | "slide_up"
  | "dissolve"
  | "zoom_in"
  | "circle";

export interface SegmentConfig {
  id: string;
  sourceStart: number;
  sourceEnd: number;
  transition: SegmentTransition;
  transitionDuration?: number; // seconds, default 0.3
}

export interface CaptionOverride {
  text?: string;
  hidden?: boolean;
  highlight?: boolean;
}

export interface EditConfig {
  outputRatio: OutputRatio;
  segments: SegmentConfig[];
  cuts: CutConfig[];
  zooms: ZoomConfig[];
  reframing: ReframingConfig;
  captions: CaptionConfig;
  transitions: TransitionConfig;
  audio: AudioConfig;
  overlays: OverlayConfig;
  captionOverrides: Record<number, CaptionOverride>;
}
